wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/ProjectTreeChangesDetector.java

```diff
@@ -0,0 +1,127 @@
+/*******************************************************************************
+ * Copyright (c) 2012-2016 Codenvy, S.A.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.api.vfs.impl.file.event.detectors;
+
+import com.google.common.annotations.Beta;
+
+import org.eclipse.che.api.core.jsonrpc.JsonRpcRequestTransmitter;
+import org.eclipse.che.api.core.jsonrpc.shared.JsonRpcRequest;
+import org.eclipse.che.api.project.shared.dto.event.FileWatcherEventType;
+import org.eclipse.che.api.project.shared.dto.event.ProjectTreeStatusUpdateDto;
+import org.eclipse.che.api.vfs.impl.file.event.EventTreeNode;
+import org.eclipse.che.api.vfs.impl.file.event.HiEvent;
+import org.eclipse.che.api.vfs.impl.file.event.HiEventDetector;
+import org.slf4j.Logger;
+
+import javax.annotation.PostConstruct;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+
+import static java.util.stream.Collectors.toSet;
+import static org.eclipse.che.dto.server.DtoFactory.newDto;
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * @author Dmitry Kuleshov
+ */
+@Beta
+@Singleton
+public class ProjectTreeChangesDetector implements HiEventDetector<ProjectTreeChangesDetector> {
+    private static final Logger LOG = getLogger(ProjectTreeChangesDetector.class);
+
+    private final JsonRpcRequestTransmitter transmitter;
+
+    private State state;
```

Comment:
Enum is used to make it more declarative, in general - looks like boolean can be used here.